### PR TITLE
Allow empty services

### LINKED ISSUE #766: toBuilder in RunConfiguration
Eclipse needs to add jvmFlags after a run config is created:

```
    jvmFlags.add("-Xdebug"); //$NON-NLS-1$
    jvmFlags.add("-Xrunjdwp:transport=dt_socket,server=n,suspend=y,quiet=y,address=" + debugPort); //$NON-NLS-1$
    devServerRunConfiguration.setJvmFlags(jvmFlags);
```

Also we have cases where a RunConfiguration does not have deployables.

### PR DESCRIPTION
**Pull request by elharo**
Continuing from #767 

fix #766

### CODE REVIEW

File: src/main/java/com/google/cloud/tools/appengine/configuration/RunConfiguration.java [hunk 1]
```diff
@@ -179,9 +179,14 @@ public String getPythonStartupArgs() {
     return pythonStartupArgs;
   }
 
-  @Nullable
+  /**
+   * Returns a mutable list that exposes the internal state.
```
Comment (suztomo) [inline src/main/java/com/google/cloud/tools/appengine/configuration/RunConfiguration.java]:
```suggestion
   * Returns JVM flags as a mutable list that exposes the internal state.
```

But I believe the implementation does not expose the mutable internal state any longer.

File: src/main/java/com/google/cloud/tools/appengine/configuration/RunConfiguration.java [hunk 2]
```diff
@@ -455,4 +463,41 @@ public RunConfiguration build() {
           projectId);
     }
   }
+
+  /**
+   * Returns a mutable builder initialized with the values of this runtime configuration.
+   *
+   * @return a mutable builder initialized with the values of this runtime configuration
```
Comment (suztomo) [inline src/main/java/com/google/cloud/tools/appengine/configuration/RunConfiguration.java]: Google-style Javadoc only needs a summary fragment. No need to maintain `@return`.
https://google.github.io/styleguide/javaguide.html#s7.2-summary-fragment

File: src/main/java/com/google/cloud/tools/appengine/configuration/RunConfiguration.java (hunk 1 above)
Comment (elharo) [reply src/main/java/com/google/cloud/tools/appengine/configuration/RunConfiguration.java]: done

File: src/main/java/com/google/cloud/tools/appengine/configuration/RunConfiguration.java (hunk 2 above)
Comment (elharo) [reply src/main/java/com/google/cloud/tools/appengine/configuration/RunConfiguration.java]: done

**Comment by chanseokoh** [comment]
Looks like the new tests are failing on Kokoro Windows.

```
[ERROR] Failures: 
[ERROR]   RunConfigurationTest.testBuilder:109 expected:<[/]storagePath> but was:<[\]storagePath>
[ERROR]   RunConfigurationTest.testToBuilder:142 expected:<[/]storagePath> but was:<[\]storagePath>
```

File: src/test/java/com/google/cloud/tools/appengine/configuration/RunConfigurationTest.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+            .runtime("runtime")
+            .skipSdkUpdateCheck(true)
+            .storagePath(Paths.get(File.separator + "storagePath"))
+            .threadsafeOverride("threadsafeOverride")
+            .useMtimeFileWatcher(true)
+            .build();
+  }
+
+  @Test
+  public void testJvmFlags() {
```
Comment (chanseokoh) [inline src/test/java/com/google/cloud/tools/appengine/configuration/RunConfigurationTest.java]: `testGetJvmFlags_returnsCopy()`

**Comment by loosebazooka** [comment]
Just curious if we're using `toBuilder` do will still need to make JvmFlags non-nullable?

**Comment by chanseokoh** [comment]
> Just curious if we're using `toBuilder` do will still need to make JvmFlags non-nullable?

Ah, that's a good point. Probably not needed.

File: pom.xml [hunk 4]
```diff
@@ -149,7 +149,7 @@
...
+            <arg>-Xplugin:ErrorProne -Xep:NullAway:ERROR -XepOpt:NullAway:AnnotatedPackages=com.google.cloud.tools -XepOpt:NullAway:KnownInitializers=com.google.gson.Gson.fromJson -XepOpt:NullAway:ExcludedFieldAnnotations=org.mockito.Mock,org.junit.runners.Parameterized -XepOpt:NullAway:ExcludedClasses=com.google.cloud.tools.appengine.operations.DevServerV1Test,com.google.cloud.tools.appengine.operations.cloudsdk.serialization.AppEngineDeployResult,com.google.cloud.tools.io.FilePermissionsTest,com.google.cloud.tools.appengine.operations.DevServerV2Test,com.google.cloud.tools.managedcloudsdk.install.InstallerFactoryTest,com.google.cloud.tools.appengine.AppEngineDescriptorTest,com.google.cloud.tools.appengine.operations.cloudsdk.serialization.CloudSdkVersionTest,com.google.cloud.tools.appengine.operations.GenRepoInfoFileTest,com.google.cloud.tools.appengine.operations.DeploymentTest,com.google.cloud.tools.appengine.operations.AuthTest,com.google.cloud.tools.appengine.configuration.RunConfigurationTest</arg>
```
Comment (chanseokoh) [inline pom.xml]: Does `RunConfigurationTest` have actually something that violates the NullAway check? A quick look at the test class doesn't seem likely.

Comment (elharo) [reply pom.xml]: At one point NullAway failed here. Let me check if it still does. We may have moved passed that.

File: src/main/java/com/google/cloud/tools/appengine/configuration/RunConfiguration.java [hunk 5]
```diff
@@ -353,8 +356,12 @@ public Builder pythonStartupArgs(@Nullable String pythonStartupArgs) {
       return this;
     }
 
+    /** Sets extra flags to be passed to the Java vortual machine. */
```
[the remark this answers is not in the document]
Comment (elharo) [reply src/main/java/com/google/cloud/tools/appengine/configuration/RunConfiguration.java]: done

File: src/main/java/com/google/cloud/tools/appengine/configuration/RunConfiguration.java [hunk 6]
```diff
@@ -353,8 +356,12 @@ public Builder pythonStartupArgs(@Nullable String pythonStartupArgs) {
...
     public Builder jvmFlags(@Nullable List<String> jvmFlags) {
```
[the remark this answers is not in the document]
Comment (elharo) [reply src/main/java/com/google/cloud/tools/appengine/configuration/RunConfiguration.java]: That does make sense, but it risks breaking existing code.

File: src/main/java/com/google/cloud/tools/appengine/configuration/RunConfiguration.java [hunk 7]
```diff
@@ -455,4 +462,37 @@ public RunConfiguration build() {
           projectId);
     }
   }
+
+  /** Returns a mutable builder initialized with the values of this runtime configuration. */
+  public Builder toBuilder() {
+    Builder builder =
+        builder(this.getServices())
+            .additionalArguments(this.getAdditionalArguments())
```
[the remark this answers is not in the document]
Comment (elharo) [reply src/main/java/com/google/cloud/tools/appengine/configuration/RunConfiguration.java]: done

File: src/test/java/com/google/cloud/tools/appengine/configuration/RunConfigurationTest.java [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+import java.util.List;
+import java.util.Map;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+public class RunConfigurationTest {
+
+  private RunConfiguration configuration;
+  private List<String> inputFlags = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (elharo) [reply src/test/java/com/google/cloud/tools/appengine/configuration/RunConfigurationTest.java]: done

File: src/test/java/com/google/cloud/tools/appengine/configuration/RunConfigurationTest.java [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+  private List<Path> services = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (elharo) [reply src/test/java/com/google/cloud/tools/appengine/configuration/RunConfigurationTest.java]: done

File: src/test/java/com/google/cloud/tools/appengine/configuration/RunConfigurationTest.java (hunk 3 above)
Comment (elharo) [reply src/test/java/com/google/cloud/tools/appengine/configuration/RunConfigurationTest.java]: It's testing a bit more than that

File: src/main/java/com/google/cloud/tools/appengine/configuration/RunConfiguration.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (elharo) [reply src/main/java/com/google/cloud/tools/appengine/configuration/RunConfiguration.java]: Effective Java, 3rd edition, Item 54: Return empty collections or arrays, not nulls

**Pull request merged**